github-runner/autoscaler/main.py

```diff
@@ -0,0 +1,233 @@
+    except ValueError:
+        raise
+
+
+def update_runner_instance_count(instance_count: int):
+    """
+    Updates a Cloud Run worker pool with the specified instance count.
+    """
+
+    url = f"{CLOUDRUN_URI}?updateMask=scaling.manualInstanceCount"
+    payload = {
+        "scaling": {"scalingMode": "MANUAL", "manualInstanceCount": instance_count}
+    }
+
+    try:
+        _call_cloudrun_api("PATCH", url=url, payload=payload)
+    except ValueError:
+        raise
+
+
+def validate_signature(request):
+    """
+    Checks the validity of the webhook, using GitHub's suggested implementation.
+
+    https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries#python-example
+    """
+    signature_header = request.headers.get("x-hub-signature-256")
+    payload_body = request.data
+
+    if not signature_header:
+        logging.error("x-hub-signature-256 header is missing!")
+        return False
+
+    hash_object = hmac.new(
+        GITHUB_WEBHOOK_SECRET.encode("utf-8"),
+        msg=payload_body,
+        digestmod=hashlib.sha256,
+    )
+    expected_signature = "sha256=" + hash_object.hexdigest()
+    if not hmac.compare_digest(expected_signature, signature_header):
+        logging.error("Request signatures didn't match!")
+        return False
+
+    return True
+
+
+# --- Main Webhook Handler ---
+def github_webhook_handler(request: Request):
+    """
+    HTTP Cloud Function that handles GitHub workflow_job events for autoscaling.
+    """
+
+    # 0. Log invocation.
+    event_type = request.headers.get("X-GitHub-Event")
+    print(f"Received event type '{event_type}'")
+
+    # 1. Validate Webhook Signature
+    if not validate_signature(request):
+        return ("Invalid signature", 403)
+
+    # 2. Parse Event
+    if event_type != "workflow_job":
+        print(f"Ignoring event type '{event_type}'")
+        return ("OK", 200)
+
+    try:
+        payload = request.get_json()
+    except Exception as e:
+        return (f"Error parsing JSON payload: {e}", 400)
+
+    action = payload.get("action")
+    job = payload.get("workflow_job")
+
+    if not job:
+        return ("No 'workflow_job' found in payload.", 200)
+
+    job_id = job.get("id")
+    job_name = job.get("name")
+    job_status = job.get("status")  # 'queued', 'in_progress', 'completed'
+    print(
+        f"Received workflow_job event: Job ID {job_id}, Name '{job_name}', "
+        f"Status '{job_status}', Action '{action}'"
+    )
+
+    # 3. Handle Scaling Logic
+    # [START run_github_worker_pool_scaling_logic]
+    try:
+        current_instance_count = get_current_worker_pool_instance_count()
+    except ValueError as e:
+        return f"Could not retrieve instance count: {e}", 500
+
+    # Scale Up: If a job is queued and we have available capacity
+    if action == "queued" and job_status == "queued":
+        print(f"Job '{job_name}' is queued.")
+
+        if current_instance_count < MAX_RUNNERS:
```

Comment:
Inferior or equal here I suppose

Reply:
That would cause the max runners to be exceeded. In the current form, if the instance count matches the max runners, no more are created.